contracts/libraries/VaultLifecycleSTETH.sol

```diff
@@ -353,66 +353,74 @@ library VaultLifecycleSTETH {
 
-        // We pass in the amount of stETH we want to unwrap from wstETH
-        // Though stETH != ETH, we assume that they are equivalent here
-        // by passing in the amount of ETH we need to withdraw
-        // This assumption is fine because we will be swapping the stETH to ETH.
-        uint256 stethNeeded =
-            DSMath.max(assetBalance, amount).sub(assetBalance);
+        // Scenario 1
+        if (ethAvailable >= amount) {
+            return amount;
+        }
+
+        // We unwrap if necessary first, then do the swap
+        uint256 ethstEthSum = ethAvailable.add(stethAvailable);
+        bool hasEnoughETHForWithdrawal = ethstEthSum >= amount.sub(1); // off by 1
+
+        // Scenario 3
+        {
+            if (!hasEnoughETHForWithdrawal) {
+                uint256 stethNeededFromUnwrap = amount.sub(ethstEthSum);
+                uint256 wstAmountToUnwrap =
+                    IWSTETH(collateralToken).getWstETHByStETH(
+                        stethNeededFromUnwrap
+                    );
+
+                IWSTETH(collateralToken).unwrap(wstAmountToUnwrap);
+
+                // We do a final double check that we have enough to satisfy the amount
+                // We account for the off by 1 error
+                stethAvailable = steth.balanceOf(address(this));
+                require(
+                    ethAvailable.add(stethAvailable) >= amount.sub(2),
+                    "Unwrapping wstETH did not return sufficient stETH"
+                );
+            }
+        }
 
-        uint256 wstETHPerStETH = IWSTETH(collateralToken).tokensPerStEth();
-        uint256 amountToUnwrap = stethNeeded.mul(wstETHPerStETH).div(10**18);
+        // Scenario 2
+        // Now that we satisfied the ETH + stETH sum, we swap the stETH amounts necessary
+        // to facilitate a withdrawal
 
-        if (amountToUnwrap > 0) {
-            IWSTETH wsteth = IWSTETH(collateralToken);
-            IERC20 steth = IERC20(stethToken);
+        // This won't underflow since we already asserted that ethAvailable < amount before this
+        uint256 stEthAmountToSwap =
+            DSMath.min(amount.sub(ethAvailable), stethAvailable);
 
-            uint256 startStethBalance = steth.balanceOf(address(this));
+        steth.safeApprove(crvPool, stEthAmountToSwap);
 
-            if (stethNeeded > startStethBalance) {
-                amountToUnwrap = stethNeeded
-                    .sub(startStethBalance)
-                    .mul(wstETHPerStETH)
-                    .div(10**18);
-                // Unwrap to stETH
-                wsteth.unwrap(amountToUnwrap);
-            }
+        // CRV SWAP HERE from steth -> eth
+        // 0 = ETH, 1 = STETH
+        // We are setting 1, which is the smallest possible value for the _minAmountOut parameter
+        // However it is fine because we check that the totalETHOut >= minETHOut at the end
+        // which makes sandwich attacks not possible
```

Comment:
This is missing the fact that the minETHOut is for the entire withdrawal amount, so for instance:

Vault holds = 5 ETH, 5 stETH
I initiate a withdrawal of amount=10 ETH, minETHOut=9.9 ETH

The curve swap would only do a swap of 5 stETH so it would always revert because the minETHOut is 9.9. So to cleanest solution here is to check at the end of the output